File: backend-dotnet/Application/Queries/Posts/GetPublicPostsQueryHandler.cs
using Application.DTOs;
using Application.Interfaces;
using Microsoft.Extensions.Logging;

namespace Application.Queries.Posts
{
    public class GetPublicPostsQueryHandler : IQueryHandler<GetPublicPostsQuery, PagedResult<PostListItemDto>>
    {
        private readonly IPostRepository _postRepository;
        private readonly ICacheService _cacheService;
        private readonly ILogger<GetPublicPostsQueryHandler> _logger;

        public GetPublicPostsQueryHandler(
            IPostRepository postRepository,
            ICacheService cacheService,
            ILogger<GetPublicPostsQueryHandler> logger)
        {
            _postRepository = postRepository;
            _cacheService = cacheService;
            _logger = logger;
        }

        public async Task<PagedResult<PostListItemDto>> HandleAsync(GetPublicPostsQuery query)
        {
            var cacheKey = $"public_posts_page_{query.Page}_{query.PageSize}_{query.CurrentUserId}";
            
            var cached = await _cacheService.GetAsync<PagedResult<PostListItemDto>>(cacheKey);
            if (cached != null)
            {
                _logger.LogDebug("Public posts page {Page} retrieved from cache", query.Page);
                return cached;
            }

            var result = await _postRepository.GetPublicPostsAsync(query.CurrentUserId, query.Page, query.PageSize);

            // Cache for 15 minutes
            await _cacheService.SetAsync(cacheKey, result, TimeSpan.FromMinutes(15));

            _logger.LogInformation("Retrieved {Count} public posts for page {Page}", result.Items.Count, query.Page);
            return result;
        }
    }
}

File: backend-java/src/main/java/com/jerrygram/application/queries/posts/GetPublicPostsQueryHandler.java
package com.jerrygram.application.queries.posts;

import com.jerrygram.application.dtos.PagedResult;
import com.jerrygram.application.dtos.PostListItemDto;
import com.jerrygram.application.interfaces.ICacheService;
import com.jerrygram.application.interfaces.IQueryHandler;
import com.jerrygram.infrastructure.repositories.PostRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.time.Duration;

@Service
@RequiredArgsConstructor
@Slf4j
public class GetPublicPostsQueryHandler implements IQueryHandler<GetPublicPostsQuery, PagedResult<PostListItemDto>> {

    private final PostRepository postRepository;
    private final ICacheService cacheService;

    @Override
    public PagedResult<PostListItemDto> handle(GetPublicPostsQuery query) {
        var cacheKey = "public_posts_page_" + query.getPage() + "_" + query.getPageSize() + "_" + query.getCurrentUserId();
        
        var cached = cacheService.get(cacheKey, PagedResult.class);
        if (cached.isPresent()) {
            log.debug("Public posts page {} retrieved from cache", query.getPage());
            @SuppressWarnings("unchecked")
            PagedResult<PostListItemDto> result = (PagedResult<PostListItemDto>) cached.get();
            return result;
        }

        var result = postRepository.getPublicPosts(query.getCurrentUserId(), query.getPage(), query.getPageSize());

        // Cache for 15 minutes
        cacheService.set(cacheKey, result, Duration.ofMinutes(15));

        log.info("Retrieved {} public posts for page {}", result.getItems().size(), query.getPage());
        return result;
    }
}